src/index.ts

```diff
@@ -31,66 +32,59 @@ const mapUrlWithParsedQuery = (fn: MapUrlWithParsedQueryFn) =>
         urlHelpers.format,
     );
 
-// Note: if/when this PR is merged, this type will be available via the Node types.
-// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33997
-type ParsedUrlQueryInput = { [key: string]: unknown };
-const addQueryToParsedUrl = ({
-    queryToAppend,
+const replaceQueryInParsedUrl = ({
+    newQuery,
 }: {
-    queryToAppend: ParsedUrlQueryInput;
+    newQuery: ParsedUrlQueryInput | ((prev: ParsedUrlQuery) => ParsedUrlQueryInput);
 }): MapUrlWithParsedQueryFn => ({ parsedUrl }) => {
-    const { auth, protocol, host, hash, pathname, query: existingQuery } = parsedUrl;
-    const newQuery = { ...existingQuery, ...queryToAppend };
+    const { auth, protocol, host, hash, pathname, query: prevQuery } = parsedUrl;
     return {
         auth,
         protocol,
         host,
         hash,
         pathname,
-        query: newQuery,
+        query: newQuery instanceof Function ? newQuery(prevQuery) : newQuery,
     };
 };
 
-export const addQueryToUrl = flipCurried(
+export const replaceQueryInUrl = flipCurried(
     pipe(
-        addQueryToParsedUrl,
+        replaceQueryInParsedUrl,
         mapUrlWithParsedQuery,
     ),
 );
 
-const replaceQueryInParsedUrl = ({
-    newQuery,
+// Note: if/when this PR is merged, this type will be available via the Node types.
+// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33997
+type ParsedUrlQueryInput = { [key: string]: unknown };
+const addQueryToParsedUrl = ({
+    queryToAppend,
 }: {
-    newQuery: ParsedUrlQueryInput;
-}): MapUrlWithParsedQueryFn => ({ parsedUrl }) => {
-    const { auth, protocol, host, hash, pathname } = parsedUrl;
-    return {
-        auth,
-        protocol,
-        host,
-        hash,
-        pathname,
-        query: newQuery,
-    };
-};
+    queryToAppend: ParsedUrlQueryInput;
+}): MapUrlWithParsedQueryFn =>
+    replaceQueryInParsedUrl({
+        newQuery: existingQuery => ({ ...existingQuery, ...queryToAppend }),
```

Comment:
Realised we can implement add/append functions in terms of the replace functions. The other changes in this PR are just re-orders.